### PR TITLE
Closes #4 Authentication

### LINKED ISSUE #4: Authenticate as an Imgur user
- Has to be oauth, if we want to upload, so let's figure out a simple way to do this
- Username/password?
- Imgur might want a PIN as well
- Forget about refresh tokens for now, but it may be a good chance to talk about LiveData if we have time
- Probably do it in one Activity and pass the token through to a MainActivity via Intent... hey it's both simple AND naive. 

No dependencies :D

### PR DESCRIPTION
**Pull request by timothyolt**
closes #4

### CODE REVIEW

**Review by nukeforum — COMMENTED** [review]
Do we need to track the `.idea/` stuff?

File: app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.timothyolt.evolutionarydesign.MainActivity
+
+class AuthenticationActivity : AppCompatActivity() {
+
+    private var textView: TextView? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        textView = TextView(this)
+        setContentView(textView)
```
Comment (nukeforum) [inline app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt:19]: What is this black magic??? 😵

File: app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        val authentication = authentication(intent)
+        if (authentication == null) {
+            textView?.text = "Redirecting to Imgur..."
+            val outgoingOAuthIntent = Intent(Intent.ACTION_VIEW).apply {
+                data = Uri.parse("https://api.imgur.com/oauth2/authorize?client_id=6b1112a4f9783ad&response_type=token")
+            }
+            startActivity(outgoingOAuthIntent)
+        } else {
+            textView?.text = "Redirecting to App..."
+            // the good old anti-pattern
```
Comment (nukeforum) [inline app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt:33]: 🎉

File: app/src/test/java/com/timothyolt/evolutionarydesign/DisplayImageTest.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package com.timothyolt.evolutionarydesign
+
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.ShapeDrawable
+import kotlin.test.*
+
+class DisplayImageTest {
```
Comment (nukeforum) [inline app/src/test/java/com/timothyolt/evolutionarydesign/DisplayImageTest.kt:7]: Do we need this committed yet?

File: app/src/test/java/com/timothyolt/evolutionarydesign/ExampleUnitTest.kt [hunk 4]
```diff
@@ -1,8 +1,6 @@
 package com.timothyolt.evolutionarydesign
 
-import org.junit.Test
-
-import org.junit.Assert.*
+import kotlin.test.*
```
Comment (nukeforum) [inline app/src/test/java/com/timothyolt/evolutionarydesign/ExampleUnitTest.kt:3]: Could probably just delete this file.

**Comment by timothyolt** [comment]
> Do we need to track the `.idea/` stuff?

In general, we should ignore .idea/*.iml but track .idea/*.xml
iml's store stuff that's generated from the gradle files, so we should definitely ignore that.

XML's is a file-by-file decision. I typically ignore workspace, and modules, because that's a user's window layout stuff.

But I think it's a good idea to check in codeInsightSettings. I specifically configured this file to share, because I did some project-level autocomplete tuning that would be helpful to everyone. 

codestyles/Project.xml and codeStyles/codeStyleConfig.xml are also two I would like to share, because these keep our autoformat settings. This PR has some changes to what should be star imported.

compiler.xml gradle.xml and jarRepositories.xml are included by default, but because we are an all-gradle project, we can probably ignore these. I would want to test it first to make sure they are auto-generated on gradle import though.

File: app/src/androidTest/java/com/timothyolt/evolutionarydesign/auth/AuthenticationInstrumentedTest.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                data = Uri.parse("http://timothyolt.com/android-evolutionary-design/login#access_token=test")
+            }
+        )
+        // todo: assert login state set
+        val mainActivity = getInstrumentation().waitForMonitorWithTimeout(mainActivityMonitor, 5000)
+        assertNotNull(mainActivity)
+        mainActivity.finish()
+        scenario.onActivity { assertEquals(true, it.isFinishing) }
+    }
+}
```
[the remark this answers is not in the document]
Comment (timothyolt) [reply app/src/androidTest/java/com/timothyolt/evolutionarydesign/auth/AuthenticationInstrumentedTest.kt:51]: Geez, I'm going to have to add Ktlint on here for antibikeshedding 😅

File: app/src/test/java/com/timothyolt/evolutionarydesign/ExampleUnitTest.kt (hunk 4 above)
Comment (timothyolt) [reply app/src/test/java/com/timothyolt/evolutionarydesign/ExampleUnitTest.kt:3]: Without at least one unit test, the gradle task will not run

File: app/src/test/java/com/timothyolt/evolutionarydesign/DisplayImageTest.kt (hunk 3 above)
Comment (timothyolt) [reply app/src/test/java/com/timothyolt/evolutionarydesign/DisplayImageTest.kt:7]: Right, will do

File: app/src/androidTest/java/com/timothyolt/evolutionarydesign/ExampleInstrumentedTest.kt [hunk 6]
```diff
@@ -2,11 +2,10 @@ package com.timothyolt.evolutionarydesign
...
 import androidx.test.platform.app.InstrumentationRegistry
 import androidx.test.ext.junit.runners.AndroidJUnit4
-
-import org.junit.Test
 import org.junit.runner.RunWith
-
-import org.junit.Assert.*
+import java.net.HttpURLConnection
+import java.net.URL
+import kotlin.test.*
```
[the remark this answers is not in the document]
Comment (timothyolt) [reply app/src/androidTest/java/com/timothyolt/evolutionarydesign/ExampleInstrumentedTest.kt:6]: Removed unused import, one import still changed from the migration to kotlin.test

File: app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt (hunk 1 above)
Comment (timothyolt) [reply app/src/main/java/com/timothyolt/evolutionarydesign/auth/AuthenticationActivity.kt:19]: :)

**Pull request merged**